src/callbacks_step/analysis_dg2d_parallel.jl

```diff
@@ -12,31 +12,92 @@ function calc_error_norms(func, u, t, analyzer,
-    Tuple{typeof(func), typeof(u), typeof(t), typeof(analyzer), TreeMesh{2},
-          typeof(equations), typeof(initial_condition), typeof(dg), typeof(cache),
-          typeof(cache_analysis)},
-    func, u, t, analyzer, mesh, equations, initial_condition, dg, cache, cache_analysis)
-
-  # Since the local L2 norm is already normalized and square-rooted, we need to undo this first
-  total_volume = mesh.tree.length_level_0^ndims(mesh)
-  global_l2_error = Vector(l2_error.^2 .* total_volume)
-  global_linf_error = Vector(linf_error)
-  MPI.Reduce!(global_l2_error, +, mpi_root(), mpi_comm())
-  MPI.Reduce!(global_linf_error, max, mpi_root(), mpi_comm())
+  l2_errors, linf_errors = calc_error_norms_per_element(func, u, t, analyzer,
+                                                        mesh, equations, initial_condition,
+                                                        dg, cache, cache_analysis)
+
+  nvars = length(l2_errors[1])
+  T = eltype(l2_errors)
+
+  # Convert errors from Vector of SVectors to Matrix for MPI communication
+  FT = eltype(l2_errors[1])
+  l2_errors = reshape(reinterpret(FT, l2_errors), nvars, :) |> Matrix
+  linf_errors = reshape(reinterpret(FT, linf_errors), nvars, :) |> Matrix
+
+  # Collect local error norms of all elements on root process
   if mpi_isroot()
-    l2_error   = convert(typeof(l2_error),   global_l2_error)
-    linf_error = convert(typeof(linf_error), global_linf_error)
+    global_l2_errors = zeros(eltype(l2_errors), nvars, cache.mpi_cache.n_elements_global)
+    global_linf_errors = similar(global_l2_errors)
+
+    n_elements_by_rank = parent(cache.mpi_cache.n_elements_by_rank) # convert OffsetArray to Array
+    l2_buf = MPI.VBuffer(global_l2_errors, nvars*n_elements_by_rank)
+    linf_buf = MPI.VBuffer(global_linf_errors, nvars*n_elements_by_rank)
+    MPI.Gatherv!(l2_errors, l2_buf, mpi_root(), mpi_comm())
+    MPI.Gatherv!(linf_errors, linf_buf, mpi_root(), mpi_comm())
   else
-    l2_error   = convert(typeof(l2_error),   NaN * global_l2_error)
-    linf_error = convert(typeof(linf_error), NaN * global_linf_error)
+    MPI.Gatherv!(l2_errors, nothing, mpi_root(), mpi_comm())
+    MPI.Gatherv!(linf_errors, nothing, mpi_root(), mpi_comm())
   end
 
-  l2_error = @. sqrt(l2_error / total_volume)
+  # Aggregate element error norms on root process
+  if mpi_isroot()
+    # Convert from Matrix to Vector of SVectors
+    global_l2_errors = reinterpret(T, vec(global_l2_errors))
+    global_linf_errors = reinterpret(T, vec(global_linf_errors))
+
+    # Aggregate element errors
+    # sum(global_l2_errors) does not produce the same result as in the serial case, while
+    # sum(global_l2_errors[:]) produces the same results but leads to unnecessary allocations.
+    # Therefore we use a hand-written loop here
+    l2_error = zero(T)
+    for error in global_l2_errors
+      l2_error += error
+    end
+    linf_error = reduce((x, y) -> max.(x, y), global_linf_errors)
```

Comment:
Why do you want to achieve exactly the same numbers? It's clear that floating point arithmetic is not commutative. Thus, users should expect differences in parallel.

Reply:
Exact equality in the global error norms makes parallel development & debugging much easier. For more details please refer to my general remark in the PR conversation.

Reply:
It would be great if we could turn your general comment into a comment in the source code here.

Reply:
Additionally, it would be great to state such a motivation in the first post of a PR to make it clear for everybody who is not involved in drafting this PR but whose review is requested.

Reply:
> Additionally, it would be great to state such a motivation in the first post of a PR to make it clear for everybody who is not involved in drafting this PR but whose review is requested.

Yes, that's my bad. I agree and next time we'll improve on this.

@lchristm could you please turn my comment on GH into a comment in the source code? Please feel free to amend/improve it as you see fit.

Reply:
Added comments in 30abccf.

Reply:
:+1: